File: src/Matcher.zig
const std = @import("std");
const Char = @import("char.zig").Char;
const Expression = @import("Expression.zig");
const Word = @import("Word.zig");
const Wordlist = @import("Wordlist.zig");
const log = std.log.scoped(.Matcher);

caches: []MatchCache,
layers: []Layer,
fuzz_max: usize,
words_max: usize,
match_count: usize,
index: usize,
allocator: *std.mem.Allocator,
wordlist: []*const Word,
output_buffer: [1024]u8,

const MatchCache = struct {
    const CacheClass = struct {
        words: std.ArrayList(*const Word),
        nonnull_transitions: Expression.State.Set,
        transitions: Expression.TransitionTable,
        allocator: *std.mem.Allocator,

        fn init(expression: *const Expression, allocator: *std.mem.Allocator) !CacheClass {
            return CacheClass{
                .words = std.ArrayList(*const Word).init(allocator),
                .nonnull_transitions = Expression.State.Set.initEmpty(),
                .transitions = try Expression.TransitionTable.init(1, expression.states.items.len, expression.fuzz + 1, allocator),
                .allocator = allocator,
            };
        }

        fn deinit(self: CacheClass) void {
            self.words.deinit();
            self.allocator.free(self.transitions.items);
        }

        fn transitionsSlice(self: CacheClass, index: usize) []Expression.State.Set {
            return self.transitions.stateSlice(0, index).items;
        }
    };

    const ClassesHashMap = std.array_hash_map.ArrayHashMap(Expression.TransitionTable, CacheClass, Expression.TransitionTable.hash, Expression.TransitionTable.eql, false);

    classes: ClassesHashMap,
    //classes: []CacheClass,
    //num_classes: usize,
    word_classes: []usize,
    nonnull_words: []*const Word,
    expression: *Expression,
    allocator: *std.mem.Allocator,
    wordlist: []*const Word,

    const Self = @This();

    pub fn init(expression: *Expression, wordlist: []*const Word) !Self {
        var allocator = expression.allocator;
        var timer = try std.time.Timer.start();

        var word_classes = try allocator.alloc(usize, wordlist.len);
        errdefer allocator.free(word_classes);

        var nonnull_words = std.ArrayList(*const Word).init(allocator);
        errdefer nonnull_words.deinit();

        var self: MatchCache = .{
            .classes = ClassesHashMap.init(allocator),
            .word_classes = word_classes,
            .expression = expression,
            .allocator = allocator,
            .nonnull_words = &[0]*const Word{},
            .wordlist = wordlist,
        };
        errdefer self.classes.deinit();

        // The first class is always the empty class
        var empty_class = try CacheClass.init(expression, allocator);
        try self.classes.put(empty_class.transitions, empty_class);

        var word_len_max: usize = 1;
        for (wordlist) |word| {
            word_len_max = std.math.max(word_len_max, word.chars.len + 1);
        }

        var transition_table = try Expression.TransitionTable.init(word_len_max, expression.states.items.len, expression.fuzz + 1, allocator);
        defer transition_table.free(allocator);

        var previous_chars: []const Char = &[0]Char{};
        var total_prefixed: usize = 0;
        var total_matched: usize = 0;
        var total_length: usize = 0;

        for (wordlist) |word, w| {
            var si: usize = 0;
            while (si < word.chars.len and si < previous_chars.len and word.chars[si] == previous_chars[si]) : (si += 1) {}

            const prefixed_table = transition_table.charSliceRange(si);
            if (si == 0) {
                expression.initTransitionTable(prefixed_table);
            }
            const valid_len = expression.fillTransitionTable(word.chars[si..], prefixed_table) + si;

            total_prefixed += si;
            total_matched += valid_len;
            total_length += word.chars.len;

            previous_chars = word.chars[0..valid_len];

            if (valid_len < word.chars.len) {
                continue;
            }
            std.debug.assert(valid_len == word.chars.len);

            const word_transitions = transition_table.charSlice(word.chars.len);
            var result = try self.classes.getOrPut(word_transitions);
            if (!result.found_existing) {
                result.entry.value = try CacheClass.init(expression, allocator);
                var class = &result.entry.value;
                std.debug.assert(class.transitions.items.len == word_transitions.items.len);
                std.mem.copy(Expression.State.Set, class.transitions.items, word_transitions.items);
                result.entry.key = class.transitions;

                for (expression.states.items) |state, i| {
                    if (!class.transitionsSlice(i)[0].isEmpty()) {
                        class.nonnull_transitions.set(i);
                    }
                }

                const c = self.classes.count();
                if (c < 20) {
                    log.info("{}: nonnull: {a}: {s}", .{ c - 1, class.nonnull_transitions, word.text });
                    //log.info("{a}", .{transition_table.charSlice(0)});
                    //log.info("{a}", .{word_transitions});
                }
            }
            try result.entry.value.words.append(word);

            if (result.index != 0) {
                self.word_classes[nonnull_words.items.len] = result.index;
                try nonnull_words.append(word);
            }
        }

        self.nonnull_words = nonnull_words.toOwnedSlice();

        const num_classes = self.classes.count();
        const dt = timer.read();
        std.debug.print("prefixed={}, matched={}, total_len={} (saved {}/word on prefix and {}/word on early term)\n", .{ total_prefixed, total_matched, total_length, total_prefixed / wordlist.len, (total_length - total_matched) / wordlist.len });
        std.debug.print("{} distinct classes with {} words in {}ms ({} ns/word)\n", .{ num_classes, self.nonnull_words.len, dt / 1_000_000, dt / wordlist.len });
        //std.debug.print("{} input words; {} nonmatch\n", .{wordlist.words.items.len, self.classes[0].words.items.len});

        return self;
    }

    pub fn reduceWordlist(self: *Self, new_wordlist: []*const Word) !void {
        //std.debug.assert(self.nonnull_words != null);
        var new_word_classes = try self.allocator.alloc(usize, new_wordlist.len);
        errdefer self.allocator.free(new_word_classes);

        var i: usize = 0;
        for (self.nonnull_words) |word, w| {
            if (i < new_wordlist.len and word == new_wordlist[i]) {
                new_word_classes[i] = self.word_classes[w];
                i += 1;
            }
        }
        std.debug.assert(i == new_wordlist.len);

        self.allocator.free(self.nonnull_words);
        self.nonnull_words = &[0]*const Word{};

        self.allocator.free(self.word_classes);
        self.word_classes = new_word_classes;
    }

    pub fn deinit(self: *Self) void {
        self.allocator.free(self.nonnull_words);
        self.allocator.free(self.word_classes);
        var it = self.classes.iterator();
        while (it.next()) |entry| {
            entry.value.deinit();
        }
        self.classes.deinit();
    }
};

const Layer = struct {
    wi: usize,
    stem: *const Word,
    states: Expression.TransitionTable,
};

pub fn init(expressions: []*Expression, wordlist: Wordlist, words_max: usize, allocator: *std.mem.Allocator) !@This() {
    var caches = try std.ArrayList(MatchCache).initCapacity(allocator, expressions.len);
    errdefer caches.deinit();

    var words = wordlist.pointer_slice;
    errdefer for (caches.items) |*cache| cache.deinit();
    for (expressions) |expr, i| {
        caches.appendAssumeCapacity(try MatchCache.init(expr, words));
        words = caches.items[i].nonnull_words;
    }

    for (caches.items[0 .. caches.items.len - 1]) |*cache| {
        try cache.reduceWordlist(words);
    }

    var fuzz_max: usize = 0;
    for (caches.items) |*cache| {
        fuzz_max = std.math.max(fuzz_max, cache.expression.fuzz);
    }
    log.info("fuzz_max={}", .{fuzz_max});

    var layers = try std.ArrayList(Layer).initCapacity(allocator, words_max + 2);
    errdefer layers.deinit();

    errdefer for (layers.items) |*layer| layer.states.free(allocator);
    var l: usize = 0;
    while (l < layers.capacity) : (l += 1) {
        layers.addOneAssumeCapacity().states = try Expression.TransitionTable.init(1, caches.items.len, fuzz_max + 1, allocator);
    }

    layers.items[0].states.clear();
    for (expressions) |expr, i| {
        expr.initTransitionTable(layers.items[0].states.stateSlice(0, i));
        //expr.matchPartial(&.{}, 0, layers.items[0].states.stateSlice(fuzz_max + 1, i));
    }
    layers.items[0].wi = 0;

    return @This(){
        .caches = caches.toOwnedSlice(),
        .layers = layers.toOwnedSlice(),
        .fuzz_max = fuzz_max,
        .words_max = words_max,
        .allocator = allocator,
        .wordlist = words,
        .output_buffer = undefined,
        .index = 0,
        .match_count = 0,
    };
}

pub fn deinit(self: *@This()) void {
    for (self.layers) |*layer| {
        layer.states.free(self.allocator);
    }
    self.allocator.free(self.layers);
    for (self.caches) |*cache| {
        cache.deinit();
    }
    self.allocator.free(self.caches);
}

fn formatOutput(self: *@This()) []const u8 {
    var writer = std.io.fixedBufferStream(&self.output_buffer);
    var i: usize = 0;
    while (i <= self.index) : (i += 1) {
        _ = writer.write(" ") catch null;
        _ = writer.write(self.layers[i].stem.text) catch null;
    }
    return writer.getWritten()[1..];
}

pub fn match(self: *@This()) ?[]const u8 {
    var result: ?[]const u8 = null;
    while (true) {
        var layer = &self.layers[self.index];
        var no_match = false;
        var all_end_match = true;
        const word = self.wordlist[layer.wi];
        match: {
            for (self.caches) |*cache, c| {
                if (cache.word_classes[layer.wi] == 0) {
                    no_match = true;
                    break :match; // non-match
                }

                var end_ss = self.layers[self.index + 1].states.stateSlice(0, c).items[0..(cache.expression.fuzz + 1)];
                var states = layer.states.stateSlice(0, c).items[0..(cache.expression.fuzz + 1)];

                for (end_ss) |*e| {
                    e.* = Expression.State.Set.initEmpty();
                }
                const class = &cache.classes.items()[cache.word_classes[layer.wi]].value;

                const verbose = false;
                //const verbose = std.mem.eql(u8, word.text, "expression") or std.mem.eql(u8, word.text, "test");
                if (verbose) {
                    log.info(">>> expr#{} word={s}, start#0={a}", .{ c, word.text, states[0] });
                }

                // This style is faster for dense (small) bitsets
                for (states) |*fuzz_states, f| {
                    var iter = fuzz_states.iterator();
                    while (iter.next()) |si| {
                        if (si >= cache.expression.states.items.len) {
                            std.debug.assert(si > 250);
                            break;
                        }
                        var fd: usize = 0;
                        while (f + fd <= cache.expression.fuzz) : (fd += 1) {
                            if (verbose) {
                                log.info("transitions from {} = {a}", .{ si, class.transitionsSlice(si)[fd] });
                            }
                            end_ss[f + fd].setUnion(class.transitionsSlice(si)[fd]);
                        }
                    }
                }

                var all_empty = true;
                var any_end_match = false;
                var success = cache.expression.states.items.len - 1;
                for (end_ss) |*es, e| {
                    if (verbose) {
                        log.info("expr#{} word={s}, start#{}={a}, es#{}={a}, success={} ({}), empty={}", .{ c, word.text, e, states[e], e, es.*, es.isSet(success), success, es.isEmpty() });
                    }
                    if (es.isSet(success)) {
                        any_end_match = true;
                        all_empty = false;
                    } else if (!es.isEmpty()) {
                        all_empty = false;
                    }
                }

                if (verbose) {
                    log.info("{s} layer={}, expr={} results: any_end_match={} all_empty={} no_match={} end_ss={a}", .{ word.text, self.index, c, any_end_match, all_empty, no_match, end_ss });
                }

                if (all_empty) {
                    std.debug.assert(!any_end_match);
                    no_match = true;
                    break;
                }
                if (!any_end_match) {
                    all_end_match = false;
                }
            }

            if (no_match) {
                break :match;
            }

            layer.stem = word;

            if (all_end_match) {
                result = self.formatOutput();
                //std.debug.print("Match: {s}\n", .{result});
                self.match_count += 1;
            }
        }
        if (self.advance(!no_match) or result != null) {
            break;
        }
    }
    return result;
}

fn advance(self: *@This(), partial_match: bool) bool {
    var pm = partial_match;
    while (true) {
        if (!pm) {
            self.layers[self.index].wi += 1;
            if (self.layers[self.index].wi >= self.wordlist.len) {
                if (self.index == 0) {
                    log.info("done", .{});
                    return true;
                }
                self.index -= 1;
                continue;
            }
            break;
        } else {
            if (self.index + 1 >= self.words_max) {
                pm = false;
                continue;
            }
            self.index += 1;
            self.layers[self.index].wi = 0;
            break;
        }
    }
    return false;
}

File: src/matcher.rs
use crate::bitset::BitSet3D;
use crate::expression::Expression;
use crate::words::*;
use indexmap::IndexMap;
use std::cell::RefCell;
use std::fmt;
use std::rc::Rc;

#[derive(Debug)]
struct CacheClass {
    n_words: usize,
}

struct CacheBuilder<'expr, 'word> {
    cache: Cache<'expr, 'word>,

    index: usize,
    wordlist: Rc<RefCell<Vec<&'word Word>>>,
    transition_table: Vec<BitSet3D>,
    previous_chars: &'word [Char],

    total_prefixed: usize,
    total_matched: usize,
    total_length: usize,
}

impl<'expr, 'word> CacheBuilder<'expr, 'word> {
    fn new(
        expression: &'expr Expression,
        wordlist: Rc<RefCell<Vec<&'word Word>>>,
        word_len_max: usize,
    ) -> Self {
        let nonnull_words = Rc::new(RefCell::new(vec![]));
        let mut classes = IndexMap::new();
        let word_classes = vec![];

        let transition_group_new = || {
            BitSet3D::new(
                (expression.states_len(), expression.fuzz + 1),
                expression.states_len(),
            )
        };

        // The first class (0) is always the "empty" class
        classes.insert_full(transition_group_new(), CacheClass { n_words: 0 });

        let transition_table = vec![transition_group_new(); word_len_max];

        Self {
            cache: Cache {
                expression,
                nonnull_words,
                classes,
                word_classes,
            },

            index: 0,
            wordlist,
            transition_table,
            previous_chars: &[],

            total_prefixed: 0,
            total_matched: 0,
            total_length: 0,
        }
    }

    /// `RUNTIME: O(words * chars * fuzz * states^3)`
    fn next_single_word(&mut self) -> Option<&'word Word> {
        // RUNTIME: O(words * chars * fuzz * states^3)
        let fuzz_range = 0..self.cache.expression.fuzz + 1;
        let wordlist = self.wordlist.borrow();
        let mut nonnull_words = self.cache.nonnull_words.borrow_mut();
        while self.index < wordlist.len() {
            let word = &wordlist[self.index];
            self.index += 1;

            // RUNTIME: O(chars)
            let word_len = word.chars.len();
            let mut si: usize = 0;
            while si < word_len
                && si < self.previous_chars.len()
                && word.chars[si] == self.previous_chars[si]
            {
                si += 1;
            }

            // RUNTIME: O(fuzz * states^2)
            let prefixed_table = &mut self.transition_table[si..];
            if si == 0 {
                // Clear table
                prefixed_table[0].borrow_mut().clear();
                for i in 0..self.cache.expression.states_len() {
                    prefixed_table[0]
                        .slice_mut((i, 0))
                        .union_with(self.cache.expression.epsilon_states(i));
                }
            }

            // RUNTIME: O(chars * fuzz * states^3)
            let valid_len = self
                .cache
                .expression
                .fill_transition_table(&word.chars[si..], prefixed_table)
                + si;

            self.total_prefixed += si;
            self.total_matched += valid_len;
            self.total_length += word_len;

            self.previous_chars = &word.chars[0..valid_len];
            if valid_len < word_len {
                continue;
            }
            assert!(valid_len == word_len);

            // RUNTIME: O(fuzz * states^2)
            let entry = self
                .cache
                .classes
                .entry(self.transition_table[word_len].clone())
                .and_modify(|v| v.n_words += 1);
            if entry.index() != 0 {
                nonnull_words.push(word);
                self.cache.word_classes.push(entry.index());
            }
            entry.or_insert(CacheClass { n_words: 1 });

            // RUNTIME: O(fuzz)
            for f in fuzz_range.clone() {
                let start_transitions = self.transition_table[word_len].slice((0, f));
                if start_transitions.contains(self.cache.expression.states_len() - 1) {
                    return Some(word);
                }
            }
        }

        None
    }

    /// `RUNTIME: O(words)`
    fn finalize(mut self, new_wordlist: Rc<RefCell<Vec<&'word Word>>>) -> Cache<'expr, 'word> {
        // Drain iterator
        (&mut self).count();

        println!(
            "prefixed={}, matched={}, elided={}",
            self.total_prefixed,
            self.total_matched - self.total_prefixed,
            self.total_length - self.total_matched
        );
        println!(
            "{} distinct classes with {}/{} words",
            self.cache.classes.len(),
            self.cache.nonnull_words.borrow().len(),
            self.wordlist.borrow().len(),
        );

        if new_wordlist != self.cache.nonnull_words {
            let new_wordlist = new_wordlist.clone();
            let new_wordlist = new_wordlist.borrow();
            let mut nonnull_words = self.cache.nonnull_words.borrow_mut();
            let mut new_word_classes = vec![];
            let mut i: usize = 0;

            // RUNTIME: O(words)
            for (&word, &class) in nonnull_words.iter().zip(self.cache.word_classes.iter()) {
                if i < new_wordlist.len() && *word == *new_wordlist[i] {
                    new_word_classes.push(class);
                    i += 1;
                }
            }
            assert!(i == new_wordlist.len());
            nonnull_words.clear();
            self.cache.word_classes = new_word_classes;
            assert!(self.cache.word_classes.len() == new_wordlist.len());
        }

        self.cache
    }
}

impl<'word> Iterator for CacheBuilder<'_, 'word> {
    type Item = &'word Word;

    fn next(&mut self) -> Option<&'word Word> {
        self.next_single_word()
    }
}

#[derive(Debug)]
struct Cache<'expr, 'word> {
    expression: &'expr Expression,
    nonnull_words: Rc<RefCell<Vec<&'word Word>>>,
    classes: IndexMap<BitSet3D, CacheClass>,
    word_classes: Vec<usize>,
}

#[derive(Debug)]
struct Layer<'word> {
    wi: usize,
    stem: Option<&'word Word>,
    states: BitSet3D,
}

impl<'word> Layer<'word> {
    fn new(caches_count: usize, fuzz_count: usize, states_count: usize) -> Self {
        Self {
            wi: 0,
            stem: None,
            states: BitSet3D::new((caches_count, fuzz_count), states_count),
        }
    }
}

pub struct Matcher<'expr, 'word> {
    expressions: Vec<&'expr Expression>,
    cache_builders: Vec<CacheBuilder<'expr, 'word>>,
    caches: Vec<Cache<'expr, 'word>>,
    layers: Vec<Layer<'word>>,
    wordlist: Rc<RefCell<Vec<&'word Word>>>,
    singles_done: bool,

    words_max: usize,

    index: usize,
    match_count: usize,
}

impl<'expr, 'word> Matcher<'expr, 'word> {
    pub fn new(
        expressions: &'expr [Expression],
        wordlist: Rc<RefCell<Vec<&'word Word>>>,
        words_max: usize,
    ) -> Self {
        assert!(!expressions.is_empty());

        let word_len_max = 1 + wordlist
            .borrow()
            .iter()
            .map(|w| w.chars.len())
            .max()
            .unwrap_or(0);

        let mut cache_builders: Vec<CacheBuilder<'_, '_>> = vec![];
        for expr in expressions {
            let words = cache_builders
                .last()
                .map(|c: &CacheBuilder| c.cache.nonnull_words.clone())
                .unwrap_or_else(|| wordlist.clone());
            let cache = CacheBuilder::new(expr, words, word_len_max);
            cache_builders.push(cache);
        }

        Matcher {
            expressions: expressions.iter().collect(),
            cache_builders,
            caches: vec![],
            singles_done: false,
            layers: vec![],
            wordlist: wordlist.clone(),
            words_max,
            index: 0,
            match_count: 0,
        }
    }

    /// `RUNTIME: O(expressions * (words + fuzz * states))`
    fn next_single(&mut self) -> Option<String> {
        assert!(!self.singles_done);

        // Check for single-word matches
        let (first_cache, remaining_caches) = self.cache_builders.split_at_mut(1);
        for word in &mut first_cache[0] {
            let mut all_match = true;
            for cache in remaining_caches.iter_mut() {
                let last_word = cache.last();
                all_match = all_match && (last_word == Some(word));
            }
            if all_match {
                return Some(word.text.clone());
            }
        }

        // Process the remaining words (even though they can't be single-word matches)
        remaining_caches.iter_mut().for_each(|c| {
            c.count();
        });

        // RUNTIME: O(expressions * words)
        let nonnull_wordlist = self
            .cache_builders
            .iter()
            .last()
            .unwrap()
            .cache
            .nonnull_words
            .clone();
        self.caches = self
            .cache_builders
            .drain(..)
            .map(|c| c.finalize(nonnull_wordlist.clone()))
            .collect();

        // RUNTIME: O(expressions * fuzz * states)
        let fuzz_max = self
            .expressions
            .iter()
            .map(|expr| expr.fuzz)
            .max()
            .unwrap_or(0);
        let states_max = self
            .expressions
            .iter()
            .map(|expr| expr.states_len())
            .max()
            .unwrap_or(1);
        let mut layers: Vec<Layer<'word>> = (0..=self.words_max)
            .map(|_| Layer::new(self.expressions.len(), fuzz_max + 1, states_max))
            .collect();

        for (i, expr) in self.expressions.iter().enumerate() {
            layers[0].states.slice2d_mut(i).clear();

            let mut starting_states = layers[0].states.slice_mut((i, 0));
            starting_states.union_with(expr.epsilon_states(0));
        }

        self.layers = layers;
        self.wordlist = nonnull_wordlist;
        self.singles_done = true;
        None
    }

    fn format_output(&self) -> String {
        self.layers[0..=self.index]
            .iter()
            .map(|layer| layer.stem.unwrap().text.as_str())
            .collect::<Vec<_>>()
            .join(" ")
    }

    /// `RUNTIME: O(words^words_max * expressions * fuzz^2 * states^2)`
    fn next_phrase(&mut self) -> Option<String> {
        assert!(self.singles_done);
        if self.wordlist.borrow().is_empty() {
            return None;
        }

        // RUNTIME: O(words^words_max * expressions * fuzz^2 * states^2)
        let mut result = None;
        loop {
            let mut no_match = false;
            {
                let wordlist = self.wordlist.borrow();
                let (lower_layers, upper_layers) = self.layers.split_at_mut(self.index + 1);
                let this_layer = &mut lower_layers[self.index];
                let next_layer = &mut upper_layers[0];

                // RUNTIME: O(expressions * fuzz^2 * states^2)
                let mut all_end_match = true;
                let mut all_no_advance = true;
                let wi = this_layer.wi;
                let word = wordlist[wi];
                for (c, cache) in self.caches.iter().enumerate() {
                    if cache.word_classes[wi] == 0 {
                        no_match = true;
                        break;
                    }

                    let class = cache.classes.get_index(cache.word_classes[wi]).unwrap().0;
                    next_layer.states.slice2d_mut(c).clear();

                    // RUNTIME: O(fuzz^2 * states^2)
                    let fuzz_limit = cache.expression.fuzz + 1;
                    for f in 0..fuzz_limit {
                        // RUNTIME: O(fuzz * states^2)
                        for si in this_layer.states.slice((c, f)).ones() {
                            // RUNTIME: O(fuzz * states)
                            let mut fd = 0;
                            while f + fd < fuzz_limit {
                                // RUNTIME: O(states)
                                next_layer
                                    .states
                                    .slice_mut((c, f + fd))
                                    .union_with(class.slice((si, fd)));
                                fd += 1;
                            }
                        }
                    }

                    // RUNTIME: O(fuzz * states)
                    let mut all_empty = true;
                    let mut all_subset = true;
                    let mut any_end_match = false;
                    let success_index = cache.expression.states_len() - 1;
                    for f in 0..fuzz_limit {
                        let es = next_layer.states.slice((c, f));
                        if es.contains(success_index) {
                            any_end_match = true;
                            all_empty = false;
                            all_subset = false;
                        } else if !es.is_empty() {
                            if !es.is_subset(&this_layer.states.slice((c, f))) {
                                all_subset = false;
                            }
                            all_empty = false;
                        }
                    }

                    if all_empty {
                        debug_assert!(!any_end_match);
                        no_match = true;
                        break;
                    }

                    if !all_subset {
                        all_no_advance = false;
                    }
                    if !any_end_match {
                        all_end_match = false;
                    }

                    // NB: This heuristic doesn't actually help that much
                    // RUNTIME: O(fuzz * states)
                    next_layer.states.compact_distance_set(c);
                }

                // Unclear if this optimization is worth it (even though it does help prevent .* blowouts)
                if all_no_advance {
                    no_match = true;
                }

                if !no_match {
                    this_layer.stem = Some(word);
                    if all_end_match && self.index >= 1 {
                        result = Some(self.format_output());
                        self.match_count += 1;
                    }
                }
            }

            if self.advance(!no_match) || result.is_some() {
                break;
            }
        }

        result
    }

    /// `RUNTIME: O(max_words)`
    fn advance(&mut self, partial_match: bool) -> bool {
        let mut pm = partial_match;
        loop {
            if !pm {
                self.layers[self.index].wi += 1;
                if self.layers[self.index].wi >= self.wordlist.borrow().len() {
                    if self.index == 0 {
                        println!(
                            "Matcher done with {} results (up to {} words)",
                            self.match_count, self.words_max
                        );
                        return true;
                    }
                    self.index -= 1;
                    continue;
                }
                break;
            } else {
                if self.index + 1 >= self.words_max {
                    pm = false;
                    continue;
                }
                self.index += 1;
                self.layers[self.index].wi = 0;
                break;
            }
        }
        false
    }
}

impl Iterator for Matcher<'_, '_> {
    type Item = String;

    fn next(&mut self) -> Option<String> {
        if !self.singles_done {
            self.next_single().or_else(|| self.next_phrase())
        } else {
            self.next_phrase()
        }
    }
}

impl<'expr, 'word> fmt::Debug for Matcher<'expr, 'word> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Matcher for {} expressions", self.caches.len())
    }
}
